dkg-gadget/src/worker.rs

```diff
@@ -478,42 +493,168 @@ where
+
+					offchain.remove(STORAGE_PREFIX, SUBMIT_KEYS_AT);
+
+					self.dkg_state.listening_for_pub_key = false;
+					self.aggregated_public_keys.keys_and_signatures = vec![];
+				}
+			}
+		}
+	}
+
+	fn gossip_public_key(&mut self, public_key: Vec<u8>, round_id: RoundId) {
+		let public = self
+			.key_store
+			.authority_id(&self.key_store.public_keys().unwrap())
+			.unwrap_or_else(|| panic!("Halp"));
+
+		if let Ok(signature) = self.key_store.sign(&public, &public_key) {
+			let encoded_signature = signature.encode();
+			let message = DKGMessage::<AuthorityId, DKGPayloadKey> {
+				id: public.clone(),
+				round_id,
+				payload: DKGMsgPayload::PublicKeyBroadcast(DKGPublicKeyMessage {
+					pub_key: public_key.clone(),
+					signature: encoded_signature.clone(),
+				}),
+			};
+
+			self.gossip_engine
+				.lock()
+				.gossip_message(dkg_topic::<B>(), message.encode(), true);
+
+			self.aggregated_public_keys
+				.keys_and_signatures
+				.push((public_key.clone(), encoded_signature));
+			info!(target: "dkg", "queued_keygen: gossiping local node  {:?} public key and signature", public)
+		}
+		else {
+			error!(target: "dkg", "queued_keygen: Could not sign public key");
+		}
+	}
+
+	fn handle_public_key_broadcast(&mut self, dkg_msg: DKGMessage<Public, DKGPayloadKey>) {
+		if !self.dkg_state.listening_for_pub_key {
+			return
+		}
+
+		match dkg_msg.payload {
+			DKGMsgPayload::PublicKeyBroadcast(msg) => {
+				let key_and_sig = (msg.pub_key, msg.signature);
+				if !self.aggregated_public_keys.keys_and_signatures.contains(&key_and_sig) {
+					self.aggregated_public_keys.keys_and_signatures.push(key_and_sig);
+
+					if let Some(latest_header) = self.latest_header.as_ref() {
+						let threshold = self.get_threshold(latest_header).unwrap() as usize;
+						let num_authorities = self.queued_validator_set.authorities.len();
+						let threshold_buffer = num_authorities.saturating_sub(threshold) / 2;
+						if self.aggregated_public_keys.keys_and_signatures.len() >=
+							(threshold + threshold_buffer)
+						{
+							let offchain = self.backend.offchain_storage();
+							self.dkg_state.listening_for_pub_key = false;
+
+							if let Some(mut offchain) = offchain {
+								offchain.set(
+									STORAGE_PREFIX,
+									AGGREGATED_PUBLIC_KEYS,
+									&self.aggregated_public_keys.encode(),
+								);
+								let submit_at = self.generate_random_delay();
+								if let Some(submit_at) = submit_at {
+									offchain.set(
+										STORAGE_PREFIX,
+										SUBMIT_KEYS_AT,
+										&submit_at.encode(),
+									);
+								}
+
+								info!(target: "dkg", "queued_keygen: Setting offchain storage keys {:?}, delay: {:?}", self.aggregated_public_keys, submit_at);
+
+								self.aggregated_public_keys.keys_and_signatures = vec![];
+							}
+						}
+					}
+				}
+			},
+			_ => {},
+		}
+	}
+
+	fn generate_random_delay(&self) -> Option<<B::Header as Header>::Number> {
```

Comment:
Can we document this? What distribution does it follow / what range of values does the delay take?